### PR TITLE
!!! FEATURE: Faster frontend routing

### PR DESCRIPTION
**Pull request by bwaidelich**
Moves most of the "expensive" logic from the `LinkingService` to the
`FrontendNodeRoutePartHandler` where it will be cached.

This leads to a great performance improvement because:
* The "current Site" is only determined once, not for every URI
* Previously we had *three* routes for the frontend, now it's
  a single one
* Shortcut resolution will happen in the handler, so it will
  be cached

Most of these performance improvements even hold true for uncached requests.

### Breaking Change

This might be a breaking change if you...:
* ...extend the `FrontendNodeRoutePartHandler` and rely on it's inner workings (for example
  with shortcut nodes¹).
  No methods were removed though, so chances are that you are not affected
* ...created (functional) tests that cover the Neos routing behavior.
  The `FrontendNodeRoutePartHandler` expects the `requestUriHost` routing
  parameter to be set (usually done via HTTP Middleware). You can simulate that behavior via:
  `$httpRequest = $httpRequest->withAttribute(ServerRequestAttributes::ROUTING_PARAMETERS, RouteParameters::createEmpty()->withParameter('requestUriHost', 'some-host'));`
* ...rely on the *internal* behavior of the `resolveShortcuts` argument of the `LinkingService`
  This is quite unlikely since we don't use that route part handler for the
  backend, where shortcut nodes are never resolved anyways.

---
¹ The Node instance passed to `resolveValue()` will no longer be the *target node* of a shortcut
  but the actual shortcut node itself. It is resolved to the target node/URI via `resolveShortcutNode()`.
  This method is `protected` so that it can be called from any subclass if needed.

Related: #3020

### CODE REVIEW

File: Neos.Neos/Classes/Domain/Repository/SiteRepository.php [hunk 1]
```diff
@@ -71,6 +71,23 @@ public function findFirstOnline()
         return $this->findOnline()->getFirst();
     }
 
+
+    /**
+     * @param string $nodeName
+     * @return Site|null
+     */
+    public function findOneByNodeName(string $nodeName): ?Site
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Domain/Repository/SiteRepository.php:79]: This is not directly related but it allows us to mock this call easier and the IDE likes this :)

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 2]
```diff
@@ -75,6 +86,17 @@ class FrontendNodeRoutePartHandler extends DynamicRoutePart implements FrontendN
...
+    /**
+     * @Flow\Inject
+     * @var NodeShortcutResolver
+     */
+    protected $nodeShortcutResolver;
+
+    /**
+     * @var Site[] indexed by the corresponding host name
+     */
+    protected $siteByHostRuntimeCache = [];
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:98]: This handler is a prototype (and it has to be since it is stateful) but the routing framework will re-use the same instance for the same route.
So 100 FE URLs on one page would previously resolve the current site 100 times

Comment (kitsunet) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:97]: Yep, cool, that makes sense

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 3]
```diff
@@ -112,7 +133,7 @@ protected function findValueToMatch($requestPath)
      * in time the route part handler is invoked, the security framework is not yet fully initialized.
      *
      * @param string $requestPath The request path (without leading "/", relative to the current Site Node)
-     * @return boolean true if the $requestPath could be matched, otherwise false
+     * @return bool|MatchResult
```
Comment (kitsunet) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:136]: I guess this (the bool for false) is for backwards compatibility? I would otherwise see something like an EmptyMatchResult for that case in the future?

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 4]
```diff
@@ -182,6 +201,13 @@ protected function convertRequestPathToNode($requestPath)
         if ($requestPathWithoutContext === '') {
             $node = $siteNode;
         } else {
+            if (!empty($this->options['uriPathSuffix'])) {
```
Comment (kitsunet) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:204]: I would probably want to extract that into a separate method but can be done later ofc.

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 5]
```diff
@@ -238,11 +265,92 @@ protected function resolveValue($node)
...
+     * @return UriConstraints
+     * @throws Exception\NoSiteException
+     */
+    protected function buildUriConstraintsForResolvedNode(NodeInterface $node): UriConstraints
+    {
+        $uriPath = $this->resolveRoutePathForNode($node);
+        $uriConstraints = UriConstraints::create()->withPath($uriPath);
+        $requestSite = $this->getCurrentSite();
+        if (!NodePaths::isSubPathOf(SiteService::SITES_ROOT_PATH, $node->getPath())) {
+            throw new Exception\NoSiteException(sprintf('The node at path "%s" is not located underneath the sites root path "%s"', $node->getPath(), SiteService::SITES_ROOT_PATH), 1604922914);
```
Comment (kitsunet) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:320]: Should this throw? because you could very well want to link to a node in another site? I think it should do that (like in the removed code from linking service)

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php [hunk 6]
```diff
@@ -574,19 +682,51 @@ protected function parseDimensionsAndNodePathFromRequestPathAllowingNonUniqueSeg
...
+     * @return Site
+     * @throws Exception\NoSiteException
+     */
+    protected function getCurrentSite(): Site
+    {
+        $requestUriHost = $this->parameters->getValue('requestUriHost');
+        if (!is_string($requestUriHost)) {
+            throw new Exception\NoSiteException('Failed to determine current site because the "requestUriHost" Routing parameter is not set', 1604860219);
+        }
+        if (!array_key_exists($requestUriHost, $this->siteByHostRuntimeCache)) {
```
Comment (kitsunet) [inline Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:712]: inversion of condition and early return would make this more readable and shave one level of indent off I think.

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: Yes for b/c but not related to this change but to the original feature in Flow (https://github.com/neos/flow-development-collection/pull/1126). I added a doc comment to make it clearer

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 4 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php]: good idea, extracted it

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 5 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:334]: The exception is caught in `resolveValue()` and turned into a corresponding log entry (like we did it for many other parts in the handler). I added a doc comment to the `@throws` annotation

File: Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php (hunk 6 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Routing/FrontendNodeRoutePartHandler.php:726]: good call, I extracted the actual site resolution to a dedicated method

**Review by kitsunet — APPROVED** [review]

**Comment by markusguenther** [comment]
Impressive changes 🙇 👍

**Comment by markusguenther** [comment]
Tested that on a local instance and it seems to work fine except the Unittests. But I am not that deep into that topic.
Good to have @kitsunet review 👍

**Comment by markusguenther** [comment]
🥳
